settings_ui: Fix scrollbar breaking when UI font size changes

dinocosta opened the pull request:
When the UI font size changes, the settings window's scrollbar would break because list item measurements were cached and not invalidated.

We're using `ListState.measure_all()` to pre-measure all items for an accurate scrollbar. When font size changes, these cached measurements become stale but weren't being refreshed.

This commit fixes the issue by calling `ListState.measure_all()` when settings change, ensuring items are re-measured on the next layout without affecting scroll position.

Closes #43683

Release Notes:

- Fixed bug where changing the UI Font Size in the Settings UI would break the scrollbar for the "Appearance" page

dinocosta committed:
fix(settings_ui): fix scrollbar breaking when ui font size changes

When the UI font size changes, the settings window's scrollbar would
break because list item measurements were cached and not invalidated.

We're using `ListState.measure_all()` to pre-measure all items for an
accurate scrollbar. When font size changes, these cached measurements
become stale but weren't being refreshed.

This commit fixes the issue by calling `ListState.measure_all()` when
settings change, ensuring items are re-measured on the next layout
without affecting scroll position.

Closes #43683

dinocosta commented on crates/settings_ui/src/settings_ui.rs:
```diff
@@ -1378,6 +1378,14 @@ impl SettingsWindow {
+            // instruct the `ListState` to re-measure the list items, as the
+            // list item heights may have changed depending on the new font
+            // size.
+            this.list_state.clone().measure_all();
```
Thank you for the review @Anthony-Eid !

> measure_all() is the wrong function you want to use here, because it tells the list to cache all element sizes but it doesn't clear the cache. You're going to want to use this.list_state.reset(element_count) which clears the cache.

Sorry, should have added that I did try this (by calling the `SettingsWindow.reset_list_state` method instead of `measure_all`) and I was seeing this weird behavior where, after resetting the list state, the scroll position would also reset to the top, leading to this janky experience, which I wanted to avoid ▼

https://github.com/user-attachments/assets/93dded67-fb57-469e-a154-95e29bdaa919

> Also, we shouldn't clear the catch every time a setting field is changed which is what this observe subscription would do. Instead we should check if the UI font size changed by copying the old value in the closure and reading the settings content.

Makes total sense, thank you for the suggestion. Will update accordingly 🙂

Anthony-Eid replied:
I just looked at the `list_state.reset` function definition and you're right, it also resets the scroll position. You might need to call `list_state.scroll_to` after reset with the logical_top position index. 

I personally don't like that solution, because it's a work around for an API we should support. We could add a `clear_cache` or `remeasure` function to the list that behaves like `reset`, but it doesn't reset the scroll position. This would only be allowed when the count of elements in the list doesn't change.

Anthony-Eid replied:
Your commit looks good! There might be odd behavior because we’re not recalculating the pixel offset of the logical top element. This is unlikely to be noticeable in the settings UI.

For example, if an element with a height of 10px has a scroll offset of 5px, and its updated size is less than 5px, it won’t be the top visible item anymore.

dinocosta committed:
refactor: only update list state when ui_font_size changes

dinocosta committed:
feat(gpui): introduce ListState::remeasure

Introduce a new method `ListState::remeasure` to allow clients to
invalidate cached measurements without changing the scroll position.
This is useful when item heights may have changed (e.g., font size
changes) but the number and identity of items remains the same.

dinocosta committed:
Merge branch 'main' into 43683-incorrect-scrollbar

dinocosta committed:
feat(gpui): add pending scroll fraction to list state

Preserve proportional scroll position when remeasuring list items. When
item heights change (e.g., due to font size changes), the scroll
position is now maintained relative to the item rather than jumping.

zed-zippy[bot] committed:
Autofix
